### PR TITLE
feat: Set, store, and use custom namespace and table partitions on write

### PR DESCRIPTION
**Pull request by carols10cents**
Connects to https://github.com/influxdata/idpe/issues/17266.

This stores custom templates for namespaces and tables in the database, if they're specified at namespace or table creation time (there are no catalog methods to allow for updating custom templates).

Table create now takes a table template. The table template can be created from a namespace template (and if a table is created implicitly through a write, it will use its namespace's partition template for this purpose) or from specified protobuf (and @domodwyer is working on the gRPC API that will make use of that feature).

Namespace create now optionally takes protobuf describing a partition template, and this PR hooks the `CreateNamespaceRequest` up to that. If no template is specified, the namespace gets the current partition-by-day default template. 

So if a table is created without a protobuf template in a namespace that was created without a protobuf template, the table will get the current partition-by-day default template through its namespace.

For existing namespace and table catalog records that have `NULL` for their `partition_template` columns, the `#[sqlx(default)]` will select the partition-by-day default template whenever these records are selected from the database. When we've backfilled all values in the database, the `#[sqlx(default)]` can be removed.

These commits are a touch messier than I'd like them to be; please tell me if I should clean them up more. Examples are: 

- some renames to match the what Dom did in https://github.com/influxdata/influxdb_iox/pull/7800 that I just mooshed in with another commit but really could be their own commit
- some intermediate state with the `TablePartitionTemplateOverride` added so that the namespace commit would compile that gets cleaned up in the table commit
- some intermediate changes to the `DefaultPartitionTemplate` and (application) `PartitionTemplate` types, and both those types end up being deleted by the end of the PR

I added a unit-ish type test for encoding the template types to JSON in the database, and I wanted to figure out how to write a similar test for decoding, but I can't figure out how to create a `sqlx::ValueRef` to pass in to `decode` without actually having a database :-/ Ideas extremely welcome!

### CODE REVIEW

File: data_types/src/partition_template.rs [hunk 1]
```diff
@@ -1,81 +1,196 @@
...
-});
+impl TablePartitionTemplateOverride {
+    /// When a table is being explicitly created, the creation request might have contained a
+    /// custom partition template for that table. If the custom partition template is present, use
+    /// it. Otherwise, use the namespace's partition template.
+    #[allow(dead_code)] // This will be used by the as-yet unwritten create table gRPC API.
+    fn new(
+        custom_table_template: Option<proto::PartitionTemplate>,
+        namespace_template: &NamespacePartitionTemplateOverride,
+    ) -> Self {
```
Comment (carols10cents) [inline data_types/src/partition_template.rs:83]: @domodwyer This is what I'm imagining the "create table" gRPC API will call, like:

```
let namespace = // look up namespace by name from the gRPC request
let table = repos
            .tables()
            .create(
                table_name_from_grpc_request,
                TablePartitionTemplateOverride::new(
                    possibly_specified_partition_template_proto_from_grpc_request,
                    &namespace.partition_template,
                ),
                namespace.id,
            )
            .await
            // ...
```

What do you think?

Comment (domodwyer) [reply data_types/src/partition_template.rs:83]: I think it looks great! I'm going to try and start this soon (hopefully!)

File: data_types/src/partition_template.rs [hunk 2]
```diff
@@ -1,81 +1,272 @@
...
+            )),
+        }],
+    })
+});
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use sqlx::Encode;
+
```
Comment (domodwyer) [inline data_types/src/partition_template.rs:200]: Lets add a fixture test to ensure the default derived template isn't changed, and add a note why:

```suggestion

    /// This test asserts the default derived partitioning scheme with no
    /// overrides.
    ///
    /// Changing this default during the lifetime of a cluster will cause the
    /// implicit (not overridden) partition schemes to change, potentially
    /// breaking the system invariant that a given primary keys maps to a
    /// single partition.
    ///
    /// You shouldn't be changing this!
    #[test]
    fn test_default_template_fixture() {
        let ns = NamespacePartitionTemplateOverride::default();
        let table = TablePartitionTemplateOverride::new(None, &ns);
        let got = table.parts().collect::<Vec<_>>();

        assert_matches!(got.as_slice(), [TemplatePart::TimeFormat("%Y-%m-%d")]);
    }

```

File: data_types/src/lib.rs [hunk 3]
```diff
@@ -299,11 +299,15 @@ pub struct Namespace {
     pub max_columns_per_table: i32,
     /// When this file was marked for deletion.
     pub deleted_at: Option<Timestamp>,
+    #[sqlx(default)]
```
[the remark this answers is not in the document]
Comment (carols10cents) [reply data_types/src/lib.rs:302]: I was thinking it was going to be temporary until the database is backfilled, but if you want to keep it and not backfill that works!

This is currently setting the default if the record is read out of the database and its `partition_template` column is NULL. Because we were planning to backfill, the current code in this PR will set the value to the default when the record is saved, so newly created namespaces and tables that use the default template will get non-null values in their `partition_template`. Should I change that to only save a value if the table or namespace is something other than the default?

[the remark this answers is not in the document]
Comment (carols10cents) [reply data_types/src/lib.rs:302]: > We just have to make sure (via a test and comment) that the default partitioning scheme doesn't change during the lifetime of a cluster - I think this is a reasonable expectation.

Hmmm what would you expect this test to look like?

**Comment by carols10cents** [comment]
>    Create a default (not overridden) namespace, read it back, assert the expected partitioning scheme is derived

Ok, [I've added a test for that](https://github.com/influxdata/influxdb_iox/pull/7790/commits/e9b8bb118a37d6614db249e98c9421c7bea22af3)!

>     Create a default (not overridden) table, read it back, assert the expected partitioning scheme is derived

I think [this is already covered here](https://github.com/influxdata/influxdb_iox/pull/7790/files#diff-2d583f1acd81672e054ddb2fb678ddd7a27914faff5ce3c6045752e16db2dc87R1099-R1120)? The test creates a table, asserts its partition template is the default, then calls `get_by_id` and asserts the return value from that equals the value from `create`? Am I missing something about what you'd like to see?

>    Create an overridden namespace, and create a table from it (no override), read it back and assert the expected partitioning scheme is derived

[So added!](https://github.com/influxdata/influxdb_iox/pull/7790/commits/f8997a892ce47ed029ef18fde6e13ffdf7d1ab9a)

**Comment by domodwyer** [comment]
> Am I missing something about what you'd like to see?

Nope, I just didn't see it! Thank you for adding the others 🙏

File: data_types/src/partition_template.rs [hunk 4]
```diff
@@ -1,81 +1,272 @@
...
-pub enum TemplatePart {
-    /// The value in a named column
-    Column(String),
-    /// Applies a  `strftime` format to the "time" column.
-    ///
-    /// For example, a time format of "%Y-%m-%d %H:%M:%S" will produce
-    /// partition key parts such as "2021-03-14 12:25:21" and
-    /// "2021-04-14 12:24:21"
-    TimeFormat(String),
+impl<DB> sqlx::Type<DB> for TablePartitionTemplateOverride
```
[the remark this answers is not in the document]
Comment (carols10cents) [reply data_types/src/partition_template.rs]: I'm into it! All this sqlx type junk is terrible to read, I'd rather have less of it 👍🏻 [Done](https://github.com/influxdata/influxdb_iox/pull/7790/commits/65124f75c15dd00117e7f2bf171e4c593ce6b497)!

File: data_types/src/partition_template.rs [hunk 5]
```diff
@@ -1,81 +1,272 @@
...
+            parts: vec![
+                proto::TemplatePart {
+                    part: Some(proto::template_part::Part::TagValue("region".into())),
+                },
+                proto::TemplatePart {
+                    part: Some(proto::template_part::Part::TimeFormat("year-%Y".into())),
+                },
+            ],
+        };
+        let expected_json_str = "\u{1}{\"parts\":[\
```
[the remark this answers is not in the document]
Comment (carols10cents) [reply data_types/src/partition_template.rs]: Looks like it's [the Postgres JSONB verison](https://github.com/launchbadge/sqlx/blob/253d8c9f696a3a2c7aa837b04cc93605a1376694/sqlx-postgres/src/types/json.rs#L70-L71) because I'm using `sqlx::Postgres` as the `DB`... I've [switched to using `sqlx::Sqlite`](https://github.com/influxdata/influxdb_iox/pull/7790/commits/6427a8158f1f494d01526c50f1e879b2fd7f2702) which requires a bit of extra dancing around but does not put this extra 1 in the JSON, so doesn't raise this question.

File: iox_catalog/src/lib.rs [hunk 6]
```diff
@@ -112,13 +112,27 @@ where
...
+                    repos
+                        .tables()
+                        .create(
+                            table_name,
+                            TablePartitionTemplateOverride::from(&schema.partition_template),
+                            schema.id,
+                        )
+                        .await?
+                }
+            };
```
[the remark this answers is not in the document]
Comment (carols10cents) [reply iox_catalog/src/lib.rs:135]: YOU SAID WE DIDN'T NEED CREATE_OR_GET ahhhhhhh what if we just put that back?

[the remark this answers is not in the document]
Comment (carols10cents) [reply iox_catalog/src/lib.rs:135]: > If not found, `err := create()`

Also what is this Go pseudocode in my Rust?? ;)

File: router/src/dml_handlers/partitioner.rs [hunk 7]
```diff
@@ -45,65 +44,51 @@ impl<T> Partitioned<T> {
...
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+impl Default for Partitioner {
+    fn default() -> Self {
+        Self::new()
     }
 }
```
[the remark this answers is not in the document]
Comment (carols10cents) [reply router/src/dml_handlers/partitioner.rs]: Easy peasy! [`240f8a6` (#7790)](https://github.com/influxdata/influxdb_iox/pull/7790/commits/240f8a65bd790c1d171dbe35f6c8f423dcdec51d)

File: data_types/src/partition_template.rs (hunk 2 above)
Comment (carols10cents) [reply data_types/src/partition_template.rs:168]: Thank you for writing up the reasoning! I've added this in [`0a6944a` (#7790)](https://github.com/influxdata/influxdb_iox/pull/7790/commits/0a6944a12f3e1492aa1083ab5862d32a69aa7bc7)

**Comment by carols10cents** [comment]
Status of this PR: I think there are 2 issues remaining:

1. Keeping `partition_template` nullable, aka not inserting the default value when a custom partition isn't specified -- ([previous conversation](https://github.com/influxdata/influxdb_iox/pull/7790#discussion_r1198789308)), relevant part to this problem:

>> Should I change that to only save a value if the table or namespace is something other than the default?
>
> If it's not too much trouble - I'm fairly confident that with the constructs we have, we can optimise our catalog storage costs and store only overrides, treating NULL as an implicit default.

It does seem like this is going to be a bit of trouble :( But I feel like I'm just holding sqlx wrong and missing something easy, _maybe_. I thought this was going to be a matter of removing the `unwrap_or_default`s from `NamespaceRepo::create`, that the insert would send in NULL for `None`, which I've done in the last commit I just pushed here. What I expected was that the tests would still pass but the difference would be the default value wouldn't be set in the database (it would be inferred when the derived `FromRow` saw that there wasn't a value and used the default)... but the tests are failing. 

Postgres ones fail with:

```
panicked at 'called `Result::unwrap()` on an `Err` value: SqlxError { source: ColumnDecode { index: "\"partition_template\"", source: UnexpectedNullError } }', iox_catalog/src/lib.rs:243:14
```

And sqlite ones fail with:

```
panicked at 'called `Result::unwrap()` on an `Err` value: SqlxError { source: ColumnDecode { index: "\"partition_template\"", source: Error("EOF while parsing a value", line: 1, column: 0) } }', iox_catalog/src/lib.rs:243:14
```

I've been thinking this is some sort of inference, the `INSERT` part of the query affecting how the `RETURNING` part of the query works, but now that I'm rereading this, I think this means the `#[sqlx(default)]` isn't actually working the way I thought it was. All the tests in this PR _have been inserting values into the `partition_template` column_, because all of them go through the catalog functions and that's what the catalog functions do, but really I need to write some tests that use some raw SQL to create records have a `NULL` `partition_template` (which is what existing rows in the database will have) and ensure the records behave like ones created after this PR.

I might _also_ be holding sqlx wrong.

2. Racy table create/get behavior this PR would potentially introduce ([discussion here](https://github.com/influxdata/influxdb_iox/pull/7790#discussion_r1198865048)), I need to write the helper function you suggested.

**Comment by carols10cents** [comment]
Ok, I _think_ I've fixed the remaining 2 issues; the 1st issue fixed in the 3rd-to-last commit and the 2nd issue fixed in the penultimate and ultimate commits. Phew.

**Review by domodwyer — APPROVED** [review]
I LIKE IT

File: data_types/src/partition_template.rs [hunk 8]
```diff
@@ -1,81 +1,264 @@
...
+    fn from(partition_template: proto::PartitionTemplate) -> Self {
+        Self(Some(SerializationWrapper(Arc::new(partition_template))))
+    }
+}
 
-impl NamespacePartitionTemplateOverride {
-    /// Create a new, immutable override for a namespace's partition template.
-    pub fn new(partition_template: PartitionTemplate) -> Self {
-        Self(partition_template)
+/// When a table is being created implicitly by a write, there is no possibility of a user-supplied
```
[the remark this answers is not in the document]
Comment (carols10cents) [reply data_types/src/partition_template.rs:36]: Oooh yeah, it definitely would be clearer making everything go through `TablePartitionTemplateOverride::new`. Will fix!

**Review by domodwyer — APPROVED** [review]
I REALLY LIKE IT

**Comment by domodwyer** [comment]
ITS BROKEN

**Comment by carols10cents** [comment]
I'M FIXING IT
